### PR TITLE
S3 trailing checksum

### PR DESCRIPTION
**Pull request by AllanFly120**
* add md5 integrity validation for s3 getObject in Node version >= 9.4.0;
* add content-length validator for Node and browsers;
* bugfix: request may retry after callback has been executed;

/cc @chrisradek

### CODE REVIEW

File: lib/services/s3.js [hunk 1]
```diff
@@ -479,6 +499,49 @@ AWS.util.update(AWS.S3.prototype, {
     }
   },
 
+  /**
+   * @api private
+   */
+  validateMd5TrailingChecksum: function validateMd5TrailingChecksum(resp) {
+    var response = resp.httpResponse;
+    var buffer = response.body;
```
Comment (chrisradek) [inline lib/services/s3.js]: In node.js, the response body is likely to be a stream. It might be easier in node to create a passthrough stream that generates the md5 hash and compares against the checksum. You'll know the expected content-length since it's passed in the response headers. Your solution here should still cover the browser use-case though.

File: lib/services/s3.js [hunk 2]
```diff
@@ -479,6 +499,49 @@ AWS.util.update(AWS.S3.prototype, {
...
+   * @api private
+   */
+  validateTrailingChecksum: function validateTrailingChecksum(resp) {
+    var response = resp.httpResponse;
+    if (
+      response.headers &&
+      response.headers['x-amz-transfer-encoding'] &&
+      response.headers['x-amz-transfer-encoding'].indexOf('append-') === 0
+    ) {
+      var hashAlgorithm = response.headers['x-amz-transfer-encoding'].substring('append-'.length);
```
Comment (chrisradek) [inline lib/services/s3.js]: I think we also want the fall-back of checking the request header to see which algorithm we used. This will be more important in the browser where x-amz-transfer-encoding may not be exposed. Either that or we document which headers have to be exposed in the browser, which I'm also ok with.

Comment (AllanFly120) [reply lib/services/s3.js]: I will add a note to add the exposed header to documentation along with default value in config. Since the feature is defaulted to disabled on the browser, the users who want to enable the checksum they should be able to see documentation.

File: lib/request.js [hunk 3]
```diff
@@ -595,77 +591,28 @@ AWS.Request = inherit({
...
-              lengthAccumulator.emit('end');
-              lengthAccumulator.end();
-            });
-            httpStream.pipe(lengthAccumulator).pipe(stream, { end: false });
-          } else {
             httpStream.pipe(stream);
-          }
         } else {
-
-          if (shouldCheckContentLength) {
```
Comment (chrisradek) [inline lib/request.js]: We'll still need to perform these checks for Node.js 0.8, especially if we start turning it on by default. Unfortunately 0.8.x doesn't include `Transform` in its `stream` package, so your implementation won't work for those cases. You'll likely need to resort to `data` listeners to perform your calculations if `Transform` doesn't exist.

File: lib/request.js [hunk 4]
```diff
@@ -595,77 +591,28 @@ AWS.Request = inherit({
...
-              }
-            });
-          }
-
           httpStream.on('data', function(arg) {
             stream.emit('data', arg);
           });
-          httpStream.on('end', checkContentLengthAndEmit);
+          httpStream.on('end', function() {
+            stream.emit('end');
```
Comment (chrisradek) [inline lib/request.js]: Do we still need this?

File: lib/services/s3.js [hunk 5]
```diff
@@ -479,6 +499,50 @@ AWS.util.update(AWS.S3.prototype, {
     }
   },
 
+  /**
+   * @api private
+   */
+  validateMd5TrailingChecksum: function validateMd5TrailingChecksum(resp) {
```
Comment (chrisradek) [inline lib/services/s3.js]: I know it's still a WIP, just noting you won't need this method anymore since you've moved the check closer to the IO!

File: lib/http/node-response-validator/integrityCheckerStream.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+		} else if (this.receivedLen + chunk.length > this.expectedDataLength) {
+			var newData = chunk.slice(0, this.expectedDataLength - this.receivedLen);
+			var newChecksum = chunk.slice(this.expectedDataLength - this.receivedLen);
+			this.checksum = NodeBuffer.Buffer.concat([this.checksum, newChecksum]);
+			this.hash.update(newData);
+			chunk = newData;
+		} else {
+			this.hash.update(chunk);
+		}
+		this.receivedLen += chunk ? chunk.length : 0;
```
[the remark this answers is not in the document]
Comment (AllanFly120) [reply lib/http/node-response-validator/integrityCheckerStream.js]: yes. empty md5 hashing would be '1B2M2Y8AsgTpgAmY7PhCfg=='

File: lib/services/s3.js (hunk 1 above)
Comment (AllanFly120) [reply lib/services/s3.js]: Mark: This part will only be performed on Browser.

File: lib/config.js [hunk 7]
```diff
@@ -82,6 +83,13 @@ var PromisesDependency;
  *   @return [Boolean] whether to disable S3 body signing when using signature version `v4`.
  *     Body signing can only be disabled when using https. Defaults to `true`.
  *
+ * @!attribute s3DisableTrailingChecksum
```
[the remark this answers is not in the document]
Comment (AllanFly120) [reply lib/config.js]: As we discussed last time, can I use `downloadTrailingChecksum` config name instead? On the Node it will default to 'md5' and on browsers this config won't be set.

File: lib/http/response-validator/attachValidators.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+	stream.on('end', function() {
+		if (expectedLen != receivedLen) {
+			stream.emit('error', AWS.util.error(
+				new Error('Stream content length mismatch. Received ' +
+					receivedLen + ' of ' + expectedLen + ' bytes.'),
+				{ code: 'StreamContentLengthMismatch' })
+			);
+		}
+	});
+}
```
Comment (AllanFly120) [inline lib/http/response-validator/attachValidators.js]: indentations are still not correct. will update

File: lib/services/s3.js [hunk 9]
```diff
@@ -369,6 +370,38 @@ AWS.util.update(AWS.S3.prototype, {
...
+      'getObject',
+    ];
+    var supportedChecksumAlgorithms = [
+      'md5', null,
+    ]
+    var responseChecksumAlgorithm = req.service.config.responseChecksumAlgorithm;
+    if (responseChecksumAlgorithm === undefined) responseChecksumAlgorithm = Boolean(AWS.util.isNode()) ? 'md5' : undefined; //set default when algorithem set to undefined
+    //trailing checksum header only supported on Node version >= 0.9.4
+    if (
+      Boolean(AWS.util.isNode) &&
```
Comment (AllanFly120) [inline lib/services/s3.js]: found this when going through it. will update

File: lib/http/response-validator/validationStreams.js [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+  if (chunk.length > 0) {
+    if (this.receivedLen >= this.expectedDataLength) {
+      this.providedChecksum.fill(
+        chunk, this.providedChecksumLen, this.providedChecksumLen += chunk.length
+      );
+      chunk = null;
+    } else if (this.receivedLen + chunk.length > this.expectedDataLength) {
+      var newData = chunk.slice(0, this.expectedDataLength - this.receivedLen);
+      var newChecksum = chunk.slice(this.expectedDataLength - this.receivedLen);
+      newChecksum.copy(this.providedChecksum, this.providedChecksumLen);
```
Comment (chrisradek) [inline lib/http/response-validator/validationStreams.js]: You'll also need to increment `this.providedChecksumLen` with `newChecksum.length` in case the chunk contains some of the expected content data, and only part of the checksum.

File: lib/http/response-validator/validationStreams.js [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+      this.hash.update(newData);
+      chunk = newData;
+    } else {
+      this.hash.update(chunk);
+    }
+    this.receivedLen += chunk ? chunk.length : 0;
+  }
+  callback(null, chunk);
```
Comment (chrisradek) [inline lib/http/response-validator/validationStreams.js]: Have you tested to see what happens if the `chunk` passed in contains only the trailing checksum, but not the entire thing? I suspect this would be an edge case but will _transform continue to be called if there's still more data upstream to read, but you've already sent `null` to the downstream stream?

File: lib/http/response-validator/validationStreams.js [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+  if (
+    this.providedChecksum.length !== CHECKSUM_BYTE_LEN ||
+    this.hash.digest('base64') !== this.providedChecksum.toString('base64')
+  ) {
+    this.emit('error', AWS.util.error(
+      new Error('Response fails integrity check.'),
+      { code: 'ResponseChecksumMismatch' })
+    );
+  }
+  callback();
```
Comment (chrisradek) [inline lib/http/response-validator/validationStreams.js]: Can you just call the `callback` with the error object? I think that takes care of emitting the error if you do.

File: lib/http/response-validator/validationStreams.js [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+ */
+ContentLengthCheckerStream.prototype._flush = function (callback) {
+  if (this.receivedLen !== this.expectedLen) {
+    this.emit('error', AWS.util.error(
+      new Error('Stream content length mismatch. Received ' +
+        this.receivedLen + ' of ' + this.expectedLen + ' bytes.'),
+      { code: 'StreamContentLengthMismatch' })
+    )
+  }
+  callback();
```
Comment (chrisradek) [inline lib/http/response-validator/validationStreams.js]: Same comment as above about passing the error to the callback instead of emitting it ourselves.

File: lib/event_listeners.js [hunk 14]
```diff
@@ -282,8 +287,18 @@ AWS.EventListeners = {
...
+        // suppress errors after request has completed in callback mode
+        var stream = resp.request.httpRequest.stream;
+        if (stream && stream._didCompleteCallback) {
+          return;
+        }
         if (err.code !== 'RequestAbortedError') {
-          var errCode = err.code === 'TimeoutError' ? err.code : 'NetworkingError';
+          var errCode = [
+            'TimeoutError',
+            'UnsupportedHashingAlgorithm',
```
[the remark this answers is not in the document]
Comment (AllanFly120) [reply lib/event_listeners.js]: That makes sense.

File: lib/http/response-validator/validationStreams.js (hunk 10 above)
Comment (AllanFly120) [reply lib/http/response-validator/validationStreams.js]: Thank you for pointing this out. I will fix it and add test accordingly.

File: lib/http/response-validator/validationStreams.js (hunk 11 above)
Comment (AllanFly120) [reply lib/http/response-validator/validationStreams.js]: Do you mean the case when the first arriving chunk is checksum but the data chunks haven't arrived yet? Is this case even possible?

Comment (chrisradek) [reply lib/http/response-validator/validationStreams.js]: I mean something like the following:
1st time this is called, `chunk` contains just the object's data.
2nd time this is called, `chunk` contains 14 bytes of the trailing checksum, only.
3rd time this is called, `chunk` contains the remaining 2 bytes of the trailing checksum.

I wonder if that last transform call would happen if we already sent `null` in the 2nd call. To be fair, I'm not sure if this is a realistic case. If node buffers a minimum number of bytes (until the end of the stream) before emitting that it's ready, then this is likely overkill to consider. If it can happen, this could end up as a frustrating edge case.

File: lib/http/node.js [hunk 15]
```diff
@@ -40,10 +41,47 @@ AWS.NodeHttpClient = AWS.util.inherit({
     delete options.timeout; // timeout isn't an HTTP option
 
     var stream = http.request(options, function (httpResp) {
-      if (stream.didCallback) return;
+      if (stream && stream.didCallback) return;
```
[the remark this answers is not in the document]
Comment (AllanFly120) [reply lib/http/node.js]: I have a [unit test](https://github.com/AllanFly120/aws-sdk-js/blob/029b2b1441fd43274ed75783538c1367b3782bb6/test/node_http_client.spec.js#L785) that I mock the `http.request`. In that case the `stream` is not available.

File: lib/http/response-validator/validationStreams.js [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+    process.nextTick(function(stream, encoding) {
+      stream.emit(
+        'error',
+        AWS.util.error(
+          new Error('Cannot validate response header: ' + encoding + ', expect \'append-md5\''),
+          {code: 'UnsupportedHashingAlgorithm'}
+        )
+      );
+    }, this, transferEncoding);
+  }
```
[the remark this answers is not in the document]
Comment (AllanFly120) [reply lib/http/response-validator/validationStreams.js]: it might be inevitable. Here data may come earlier than error event get thrown and stream unpiped. Then we need to access instantiated objects to do the transform, in which case if we return too early, we may access undefined objects

[the remark this answers is not in the document]
Comment (chrisradek) [reply lib/http/response-validator/validationStreams.js]: I think it makes sense to just throw the error immediately - since this is a constructor method and the condition we're checking is synchronous - rather than emit an error. This way we'll fail fast and not waste resources. It's also what other node APIs, like fs.createReadStream do if you pass in bad parameters.

File: lib/http/response-validator/integrityChecker.js [hunk 17]
```diff
@@ -53,10 +48,10 @@ IntegrityCheckerStream.prototype._flush = function(callback) {
     this.providedChecksum.length !== CHECKSUM_BYTE_LEN ||
     this.hash.digest('base64') !== this.providedChecksum.toString('base64')
   ) {
-    this.emit('error', AWS.util.error(
+    callback( AWS.util.error(
       new Error('Response fails integrity check.'),
       { code: 'ResponseChecksumMismatch' })
-    );
+    )
```
Comment (chrisradek) [inline lib/http/response-validator/integrityChecker.js]: Call `return` either on the same line as `callback` or right under. Right now `callback` might be triggered twice: once with an error and then once without an error a couple lines below.

Comment (AllanFly120) [reply lib/http/response-validator/integrityChecker.js]: Here calling callback twice won't actually write anything to stream. I will fix it though

File: test/integration_test/helpers.js [hunk 18]
```diff
@@ -0,0 +1,25 @@
+(function() {
+  global.expect = require('chai').expect;
+  module.exports = {
+    sharedBucket: 'aws-sdk-js-integration',
```
Comment (chrisradek) [inline test/integration_test/helpers.js]: Probably want to append a timestamp to this as well to make it somewhat unique

File: test/integration_test/s3/getObject.spec.js [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+    }).then(function() {
+      if (typeof done === 'function') done();
+    }).catch(function(err) {
+      throw new Error('Cannot delete bucket: ' + err);
+      exit(1);
+    })
+  }
+
+  before('setup s3 client with responseChecksumAlgorithm equals \'md5\'', function() {
+    s3 = new AWS.S3({responseChecksumAlgorithm: 'md5'});
```
Comment (chrisradek) [inline test/integration_test/s3/getObject.spec.js]: Why don't you create the bucket used by all the tests in the `before` step? We shouldn't have to create a new bucket for every single test, just this suite of tests. 

I also wouldn't mix the `putObject` method with `createBucket`. `putObject` is doing too much, and adds 'global' (across tests) state. For example, you don't directly pass it the bucket or key, instead relying on a closure that every test has access to (and can change). That could lead to tricky edge cases coming up later that are hard to debug.

File: test/integration_test/s3/getObject.spec.js [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+
+  beforeEach('setup bucket and object...', function(done) {
+    params = {
+      Bucket: bucketName,
+      Key: 'key',
+    };
+    putObject({Body: 'this is a test!'}, done);
+  })
+
+  afterEach('delete bucket...', function(done) {
```
Comment (chrisradek) [inline test/integration_test/s3/getObject.spec.js]: If you `createBucket` in the `before` hook, you can `deleteBucket` in the `after` hook!

File: test/integration_test/s3/getObject.spec.js [hunk 21]
```diff
@@ -0,0 +1,209 @@
...
+    }).createReadStream();
+    s.on('readable', function() {
+      var dat = s.read();
+      if(dat) {
+        data += dat;
+      }
+    })
+    s.on('error', function(err) {
+      error = err;
+    })
```
Comment (chrisradek) [inline test/integration_test/s3/getObject.spec.js]: Nit: Missing semi-colons here and a few other places.

File: test/integration_test/s3/getObject.spec.js [hunk 22]
```diff
@@ -0,0 +1,209 @@
...
+      }
+    })
+    s.on('error', function(err) {
+      error = err;
+      expect(error).to.not.be['null'];
+      expect(error.code).to.equal('NoSuchKey');
+      done();
+    })
+    s.on('end', function() {
+      expect(data).to.equal('');
```
Comment (chrisradek) [inline test/integration_test/s3/getObject.spec.js]: Does `end` ever get triggered? If it shouldn't, maybe we should throw an error here. If it should, then `done()` should land here assuming `end` is called after `error`.

File: test/integration_test/helpers.js [hunk 23]
```diff
@@ -0,0 +1,25 @@
...
+    AWS: require('../../lib/aws'),
+    createBuffer: function(size) {
```
Comment (chrisradek) [inline test/integration_test/helpers.js]: This method makes sense in cucumber where you are supposed to write 'human' parseable tests, but I think it's overkill for mocha. You could have constants for 1KB, 1MB if you want, then call `new Buffer` in the test. 

Then there's less code that can possibly go wrong :wink:

File: test/integration_test/helpers.js [hunk 24]
```diff
@@ -0,0 +1,25 @@
+(function() {
+  global.expect = require('chai').expect;
```
Comment (chrisradek) [inline test/integration_test/helpers.js]: Actually, if you get rid of the createBuffer method, there's really no need for this file. You could add `var expect = require('chai').expect;` to the top of your integration test file, and import the SDK directly in your test as well.

File: test/integration_test/s3/getObject.spec.js [hunk 25]
```diff
@@ -0,0 +1,209 @@
...
+    }).then(function() {
+      return s3.waitFor('objectExists', {
+        Bucket: params.Bucket,
+        Key: params.Key,
+      }).promise();
+    }).then(function() { 
+      if (typeof done === 'function') done();
+    }).catch(function(err) {
+      throw new Error('Cannot put object: ' + err);
+      exit(1);
```
Comment (chrisradek) [inline test/integration_test/s3/getObject.spec.js]: Is `exit` exposed as a global? I thought it was `process.exit()`, and am not seeing `exit` in the node repl.

Also think this line is skipped since you're throwing an error right before it.

File: features/s3/buckets.feature [hunk 26]
```diff
@@ -75,27 +75,3 @@ Feature: Working with Buckets
     Then the object "largeobject" should exist in the bucket
     Then I delete the object "largeobject" from the bucket
     Then I delete the bucket
-
-  @trailingchecksum @checksum
```
Comment (chrisradek) [inline features/s3/buckets.feature]: +1000 for not using Cucumber!

File: test/integration_test/s3/getObject.spec.js (hunk 19 above)
Comment (AllanFly120) [reply test/integration_test/s3/getObject.spec.js]: Yes, you can pass in the bucket and key, but right, it will change the global state. But capsulizing them will also guarantee `putObject` always succeed(bucket always there).

Comment (chrisradek) [reply test/integration_test/s3/getObject.spec.js]: But you could also make sure the bucket is there in the `before` hook. Just call `done()` after the `waitFor` method completes. Then you also only need to create it once; I don't think there's a reason we need to create a new bucket for every test since we aren't testing any bucket-specific configurations here. Creating a new bucket with each test also creates more points of failure (rate/resource limits, for example).

Comment (AllanFly120) [reply test/integration_test/s3/getObject.spec.js]: yea. I will move the `putObject` to `before` trait. And use timestamp in bucket name. I was thinking that we can write less code if we are going to add more tests here. But now I agree that this is an overkill and not worthwhile.

File: test/integration_test/s3/getObject.spec.js [hunk 27]
```diff
@@ -1,23 +1,29 @@
...
-      return s3.putObject({
-        Bucket: params.Bucket,
-        Key: params.Key,
-        Body: customParams.Body || params.Body
-      }).promise();
-    }).then(function() {
+      if (typeof done === 'function') done();
+    }).catch(function(err) {
+      throw new Error('Cannot create bucket: ' + err);
+      process.exit(1);
```
Comment (chrisradek) [inline test/integration_test/s3/getObject.spec.js]: Does process.exit(1) ever get called? The error that's thrown right above it should cause you to break out of this catch block.

File: test/integration_test/s3/getObject.spec.js [hunk 28]
```diff
@@ -74,7 +80,11 @@ describe('download integrity', function() {
       Key: 'key',
     };
     s3 = new AWS.S3({responseChecksumAlgorithm: 'md5'});
-    putObject({Body: 'this is a test!'}, done);
+    createBucket({}, done);
```
Comment (chrisradek) [inline test/integration_test/s3/getObject.spec.js]: Can you pass the bucket name into this function? Right now I would have to look at the `createBucket` code to understand what bucket name is being chosen and how, and I don't know what the empty object is for.
